api/v1/harborcluster_types.go

```diff
@@ -16,25 +16,321 @@ limitations under the License.
+	Kind string `json:"kind,omitempty"`
+
+	// inCluster params following.
+	Server   *RedisServer `json:"server,omitempty"`
+	Sentinel *Sentinel    `json:"sentinel,omitempty"`
+
+	// External params following.
+	// The secret must contains "address:port","usernane" and "password".
+	SecretName string `json:"secretName,omitempty"`
+	// Maximum number of socket connections.
+	// Default is 10 connections per every CPU as reported by runtime.NumCPU.
+	PoolSize int `json:"poolSize,omitempty"`
+	// TLS Config to use. When set TLS will be negotiated.
+	// set the secret which type of Opaque, and contains "tls.key","tls.crt","ca.crt".
+	// optional
+	TlsConfig string `json:"tlsConfig,omitempty"`
+}
+
+type Sentinel struct {
+	Replicas int `json:"replicas,omitempty"`
+}
+
+type RedisServer struct {
+	Replicas         int                 `json:"replicas,omitempty"`
+	Resources        corev1.ResourceList `json:"resources,omitempty"`
+	StorageClassName string              `json:"storageClassName,omitempty"`
+	// the size of storage used in redis.
+	Storage string `json:"storage,omitempty"`
+}
+
+type ChartMuseum struct {
+	AbsoluteURL bool `json:"absoluteURL,omitempty"`
+}
+
+type Trivy struct {
+	GithubToken string `json:"githubToken,omitempty"`
+}
+
+type ImageSource struct {
+	Registry         string   `json:"registry,omitempty"`
+	ImagePullSecrets []string `json:"imagePullSecrets,omitempty"`
+}
+
+type Clair struct {
+	UpdateInterval       int      `json:"updateInterval,omitempty"`
+	VulnerabilitySources []string `json:"vulnerabilitySources,omitempty"`
+}
+
+type JobService struct {
+	// +kubebuilder:validation:Required
+	Replicas string `json:"replicas"`
+
+	// +optional
+	WorkerCount int32 `json:"workerCount"`
 }
 
 // HarborClusterStatus defines the observed state of HarborCluster
 type HarborClusterStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
+	Conditions []HarborClusterCondition `json:"conditions,omitempty"`
+
+	ComponentsStatus map[Component]*ComponentsStatus `json:"ComponentsStatus,omitempty"`
+}
+
+type ComponentsStatus struct {
+	Replicas     int `json:"replicas,omitempty"`
+	ReadyReplcas int `json:"readReplcas,omitempty"`
+}
+
+// HarborClusterConditionType is a valid value for HarborClusterConditionType.Type
+type HarborClusterConditionType string
+
+// HarborClusterCondition contains details for the current condition of this pod.
+type HarborClusterCondition struct {
+	// Type is the type of the condition.
+	Type HarborClusterConditionType `json:"type"`
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status"`
+	// Last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" protobuf:"bytes,3,opt,name=lastProbeTime"`
```

Comment:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties In k8s API-Conversions, It recommends to use `LastHeartbeatTime`. And 

> Use the LastHeartbeatTime with great caution - frequent changes to this field can cause a large fan-out effect for some resources.

Reply:
good point